031-custom_authentication.md

```diff
@@ -0,0 +1,198 @@
+listener.name.<listener-name>.<sasl-mechanism>.sasl.login.callback.handler.class=
+listener.name.<listener-name>.<sasl-mechanism>.connections.max.reauth.ms=
+listener.name.<listener-name>.<sasl-mechanism>.sasl.jaas.config=
+listener.name.<listener-name>.sasl.enabled.mechanisms=
+listener.name.<listener-name>.ssl.client.auth=
+listener.security.protocol.map=
+principal.builder.class=
+```
+
+- `<listener-name>` would be the name for a given listener. 
+- `<sasl-mechanism>` would be the sasl mechanism which you wish to configure. Seeing as users can configure/build their own sasl mechanisms, all values will be allowed for this field, rather than having an allowlist of the standard sasl mechanisms. 
+
+To walkthrough an actual example, let's try configuring custom authentication for an oauthbearer workflow, where we would like to set the following properties:
+
+```
+listener.name.<listener-name>.ssl.keystore.location=/tmp/kafka/cluster.keystore.p12
+listener.name.<listener-name>.ssl.keystore.password=<password>
+listener.name.<listener-name>.ssl.keystore.type=PKCS12
+listener.name.<listener-name>.oauthbearer.sasl.client.callback.handler.class=client.class
+listener.name.<listener-name>.oauthbearer.sasl.server.callback.handler.class=server.class
+listener.name.<listener-name>.oauthbearer.sasl.login.callback.handler.class=login.class
+listener.name.<listener-name>.oauthbearer.connections.max.reauth.ms=999999999
+listener.name.<listener-name>.sasl.enabled.mechanisms=oauthbearer
+listener.name.<listener-name>.oauthbearer.sasl.jaas.config=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required ;
+principal.builder.class=SimplePrincipal.class
+listener.security.protocol.map=CONTROLPLANE-9090:SSL,REPLICATION-9091:SSL,TLS-9093:SASL_SSL
+```
+
+To achieve this, we'd simply add the following listener config to our Kafka custom resource.
+
+```
+    listeners:
+      - name: custom-auth-listener
+        port: 9093
+        type: internal
+        tls: true
+        configuration:
+          brokerCertChainAndKey:
+            secretName: my-secret
+            certificate: my-certifciate.crt
+            key: my-key.key
+        authentication:
+          type: custom
+          sasl: true
+          principal.builder.class: SimplePrincipal.class
+          listener-config:
+            oauthbearer.sasl.client.callback.handler.class: client.class
+            oauthbearer.sasl.server.callback.handler.class: server.class
+            oauthbearer.sasl.login.callback.handler.class: login.class
+            oauthbearer.connections.max.reauth.ms: 999999999
+            sasl.enabled.mechanisms: oauthbearer
+            oauthbearer.sasl.jaas.config: |
+              org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required ;
+          tlsTrustedCertificates: ...
+          secrets:
+            - name: example
+              mountPath: "/etc/example"
```

Comment:
I don't think the mount path should be free to specify. I think all these secrets should be mounted into a single subdirectory (similarly to for example done in KafkaConnect with external configurations). That will make it isolated and will also simplify validation of what is valid and what is not. Otherwise you might need to validate whether the mount points don't conflict between listeners, whether there are no security issue with the selected mount path etc.

Or is there any specific value in having this as a free configuration?

Reply:
Makes sense to me, and reduces the need to configure another thing. Only value I can see is it would be nice for deduplication, as in, I have two secrets of the same type (tls), and I can just mount them into different directories to avoid a clash. That said, I think this case is rare, and the end user could just rename their secrets. Also would be easy to adapt a rename function here, but I think for now we keep it simple, especially as there is a workaround (i.e. just rename your secret key).